mdoc/src/main/scala-3/mdoc/internal/markdown/Instrumenter.scala

```diff
@@ -0,0 +1,222 @@
+package mdoc.internal.markdown
+
+import java.io.ByteArrayOutputStream
+import java.io.PrintStream
+
+import dotty.tools.dotc.interfaces.SourcePosition
+import Instrumenter.position
+import scala.meta.Mod.Lazy
+import scala.collection.mutable
+import mdoc.Reporter
+import mdoc.internal.cli.InputFile
+import java.nio.file.Path
+import mdoc.internal.cli.Settings
+import dotty.tools.dotc.ast.untpd._
+import dotty.tools.dotc.ast.Trees
+import scala.meta.Name
+
+class Instrumenter(
+    file: InputFile,
+    sections: List[SectionInput],
+    settings: Settings,
+    reporter: Reporter
+) {
+  def instrument(): Instrumented = {
+    printAsScript()
+    Instrumented.fromSource(
+      out.toString,
+      magic.scalacOptions.toList,
+      magic.dependencies.toList,
+      magic.repositories.toList,
+      Nil,
+      reporter
+    )
+  }
+
+  val magic = new MagicImports(settings, reporter, file)
+  private val out = new ByteArrayOutputStream()
+  private val sb = new PrintStream(out)
+  val gensym = new Gensym()
+  val nest = new Nesting(sb)
+
+  private def printlnWithIndent(out: String) = {
```

Comment:
Added utility methods to make sure identation is correct, which is important for Scala 3.

I also replaced the parser.